src/os.zig

```diff
@@ -0,0 +1,371 @@
+            .win = Win32.UserProfile,
+        }).query(allocator);
+    }
+};
+
+/// POSIX-specific OS functionality.
+pub const Posix = struct {
+    pub const HomeDir = struct {
+        const Self = @This();
+
+        _path: union(enum) {
+            managed: pathlib.AutoAbsPath,
+            unmanaged: []const u8,
+        },
+
+        pub fn slice(self: *const Self) []const u8 {
+            return switch (self._path) {
+                .managed => self._path.managed.slice(),
+                .unmanaged => self._path.unmanaged,
+            };
+        }
+
+        pub fn deinit(self: *const Self) void {
+            switch (self._path) {
+                .managed => return self._path.managed.deinit(),
+                .unmanaged => {},
+            }
+        }
+
+        /// Deduces the current user's home directory on POSIX systems.
+        ///
+        /// Whichever of the following returns a value is returned.
+        /// - Per doi:10.1109/IEEESTD.2018.8277153, the `$HOME` variable.
+        /// - The `getpwuid_r` function.
+        fn query(allocator: std.mem.Allocator) !Self {
+            const Heuristics = struct {
+                fn getpwuidr(alloc: std.mem.Allocator) !Self {
+                    // The maximum total number of attempts we will have at reading getpwuid_r
+                    // before giving up. There are a few cases which benefit from re-attempting a
+                    // read.
+                    const max_attempts: comptime_int = 8;
+
+                    const initial_buf_size: usize =
+                        @intCast(bun.sys.sysconf(bun.c._SC_GETPW_R_SIZE_MAX, .{
+                            .default = std.fs.max_path_bytes,
+                        }));
+                    const buf_size_gain = 4;
+
+                    var buffer_size: usize = initial_buf_size;
+                    var temp_buf = try alloc.alloc(u8, buffer_size);
+                    defer alloc.free(temp_buf);
+
+                    for (0..max_attempts) |_| {
+                        var passwd: bun.c.struct_passwd = undefined;
+                        var result: ?*bun.c.struct_passwd = undefined;
+
+                        const rc = bun.c.getpwuid_r(
+                            Uid.queryCurrent()._underlying,
+                            &passwd,
+                            temp_buf.ptr,
+                            temp_buf.len,
+                            @ptrCast(&result),
+                        );
+                        if (rc == 0) {
+                            // Great, we found a password entry, with a home directory.
+                            if (result == null) {
+                                @branchHint(.cold);
+                                @panic("getpwuid_r returned 0 but result is null. This is a " ++
+                                    "bug in Bun. Please report the issue on GitHub.");
+                            }
+
+                            if (result.?.pw_dir == null) {
```

Comment:
As discussed, would be good to restructure to `const some_result = result orelse { @panic(...) }; const pw_dir = some_result.pw_dir orelse return error.FailedToFindHomeDir;` etc.